tf/cloud-function-v2/cloud-function.tf

```diff
@@ -0,0 +1,103 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+resource "google_cloudfunctions2_function" "function" {
+  name     = "e2etest-${terraform.workspace}"
+  location = "us-central1"
+
+  timeouts {
+    create = "5m"
+  }
+
+  build_config {
+    runtime = var.runtime
+    entry_point = var.entrypoint
+    source {
+      storage_source {
+        bucket = google_storage_bucket.bucket.name
+        object = google_storage_bucket_object.object.name
+      }
+    }
+  }
+
+  service_config {
+    max_instance_count = 1
+    available_memory = "256M"
+    timeout_seconds = 540
+    environment_variables = {
+      # Environment variables available during function execution
+      "PROJECT_ID" = var.project_id
+      "RESPONSE_TOPIC_NAME" = module.pubsub.info.response_topic.topic_name
+    }
+    ingress_settings = "ALLOW_INTERNAL_ONLY"
+    all_traffic_on_latest_revision = true    
+  }
+
+  event_trigger {
+    trigger_region = "us-central1"
+    event_type = "google.cloud.pubsub.topic.v1.messagePublished"
+    pubsub_topic =  google_pubsub_topic.topic.id
+    retry_policy = "RETRY_POLICY_RETRY"    
+  }
+}
+
+resource "google_pubsub_topic" "topic" {
+  name = "cloudfunctions-push-topic"
+}
```

Comment:
We're already including the pubsub module, can we use the request topic it creates for the `event_trigger`?

Also since name is hardcoded here, independent tests (e.g. separate PRs) are not isolated because they will all be using the same topic. That's why `terraform.workspace` name is added here https://github.com/GoogleCloudPlatform/opentelemetry-operations-e2e-testing/blob/c3a1f33e5bb0a303f974fbb8a90f6e453bad7e92/tf/modules/pubsub/main.tf#L18

---

I remember we discussed last week, but what was the reason we can't handle creating the push subscription the same way as we do for Cloud Run? 

https://github.com/GoogleCloudPlatform/opentelemetry-operations-e2e-testing/blob/c3a1f33e5bb0a303f974fbb8a90f6e453bad7e92/tf/cloud-run/cloud-run.tf#L77-L83

Reply:
Good point on the naming. Will update the topic name to be one using `terraform.workspace`. 

As for the omitting `event_trigger`, I think we can omit the triggers since for Gen2 functions - all of them are assigned an HTTP endpoint (source: https://cloud.google.com/functions/docs/calling/http) regardless of the trigger type. And then we can handle them as standard HTTP requests. But I imagine this would not require handling any subscriptions on the function side. 

I did not try this though since CloudFunctions have Pub/Sub triggers built-in and using HTTP triggers which are invoked by Pub/Sub seemed a little off. 

Seeing as EventArc was an added feature in Gen2, I think we should be using that for Gen2 cloud functions. I think we will have to use the HTTP trigger for Gen1 functions since native support for EventArc is not supported for Gen1 (See differences b/w versions [here](https://cloud.google.com/functions/docs/concepts/version-comparison)).

Reply:
> But I imagine this would not require handling any subscriptions on the function side.

what do you mean by this?

Do you know if the resources created behind the scenes by the eventarc integration are deleted when the cloud function is torn down? Is there a downside to using the pubsub subscription directly, given we already have the code for it?

Reply:
> > But I imagine this would not require handling any subscriptions on the function side.
> 
> what do you mean by this?
> 

Like in Cloud-Run we have had to run a loop on the main thread while the subscriber consumes, **I think** we wont have to do that here - cloud function should be invoked on-demand. I am not a 100% sure. 

Yes, I monitored the function while it was being created, and the only "behind the scenes" resources it creates is a subscription and attaches it automatically to the topic specified for pubsub trigger. After the function is deleted as part of the scheduled cleanup, the subscription  along with the eventarc trigger is also deleted.

Reply:
> **I think** we wont have to do that here - cloud function should be invoked on-demand. I am not a 100% sure.

we are still importing the pubsub module in this file, it's creating the request and response topics+subscriptions anyway. Good to know the cleanup is working this way.

I'm ok with this approach anyway

Reply:
Right, but the code does not have to explicitly wait on/interact with those resources.